src/MonoTorrent/MonoTorrent.Client/Managers/TrackerManager.cs

```diff
@@ -108,80 +108,114 @@ internal TrackerManager (ITrackerRequestFactory requestFactory, IEnumerable<RawT
+        async Task AnnounceToTier (TorrentEvent clientEvent, TrackerTier tier)
+        {
+            for (int i = 0; i < tier.Trackers.Count; i++) {
+                int trackerIndex = (i + tier.ActiveTrackerIndex) % tier.Trackers.Count;
+                var tracker = tier.Trackers[trackerIndex];
+
+                var result = await Announce (clientEvent, tracker);
+                tier.LastAnnounce = ValueStopwatch.StartNew ();
+                if (result) {
+                    tier.ActiveTrackerIndex = trackerIndex;
+                    tier.LastAnnounceSucceeded = true;
                     return;
-                } catch {
                 }
             }
 
-            LastAnnounceSucceeded = false;
-            AnnounceComplete?.InvokeAsync (this, new AnnounceResponseEventArgs (null, false));
+            // All trackers failed to respond.
+            tier.LastAnnounceSucceeded = false;
+        }
+
+        async Task<bool> Announce (TorrentEvent clientEvent, ITracker tracker)
+        {
+            var trackerTier = Tiers.First (t => t.Trackers.Contains (tracker));
+            try {
+                // If we have not announced to this Tracker tier yet then we should replace the ClientEvent.
+                // But if we end up announcing to a different Tracker tier we may want to send the
+                // original/unmodified args.
+                AnnounceParameters actualArgs = RequestFactory.CreateAnnounce (clientEvent);
+                if (!trackerTier.SentStartedEvent)
+                    actualArgs = actualArgs.WithClientEvent (TorrentEvent.Started);
+
+                List<Peer> peers = await tracker.AnnounceAsync (actualArgs);
+                trackerTier.LastAnnounceSucceeded = true;
+
+                trackerTier.ActiveTrackerIndex = trackerTier.Trackers.IndexOf (tracker);
+                trackerTier.SentStartedEvent |= actualArgs.ClientEvent == TorrentEvent.Started;
+                trackerTier.LastAnnounce = ValueStopwatch.StartNew ();
+                AnnounceComplete?.InvokeAsync (this, new AnnounceResponseEventArgs (tracker, true, peers.AsReadOnly ()));
+                return true;
+            } catch {
+            }
+
+            trackerTier.LastAnnounceSucceeded = false;
+            AnnounceComplete?.InvokeAsync (this, new AnnounceResponseEventArgs (tracker, false));
+            return false;
         }
 
         public async Task Scrape ()
         {
-            await Scrape (null);
+            var scrapes = new List<Task> ();
+            for (int i = 0; i < Tiers.Count; i++) {
+                var tier = Tiers[i];
+                for (int j = 0; j < tier.Trackers.Count; j++) {
+                    var tracker = tier.Trackers[j];
+                    if (tracker.CanScrape && tier.TimeSinceLastScrape > tracker.UpdateInterval)
+                        scrapes.Add (Scrape (tracker));
```

Comment:
Why this logic is different than announces?
Shouldn't scrape also be per tier?

Reply:
Yeah, it should be! This will be changed soon enough!